ilastik/applets/edgeTraining/opEdgeTraining.py

```diff
@@ -387,37 +387,53 @@ class OpPredictEdgeProbabilities(Operator):
     TrainRandomForest = InputSlot(value=False)
     EdgeClassifier = InputSlot()
     EdgeFeaturesDataFrame = InputSlot()
+    # historically named Edge"Probabilities" because multicut used predictions from
+    # random forest as edge weights. With neural networks, edge probabilities are often already good
+    # enough to act as weights.
     EdgeProbabilities = OutputSlot()  # A 1D array of probabilities, in same order as EdgeFeaturesDataFrame
 
     def setupOutputs(self):
         self.EdgeProbabilities.meta.shape = (1,)
         self.EdgeProbabilities.meta.dtype = object
 
-    def execute(self, slot, subindex, roi, result):
-        pd.set_option("display.expand_frame_repr", False)
-        if self.TrainRandomForest.value:
-            edge_features_df = self.EdgeFeaturesDataFrame.value
-            classifier = self.EdgeClassifier.value
+    def _edge_weights_from_random_forest_predictions(self):
+        edge_features_df = self.EdgeFeaturesDataFrame.value
+        classifier = self.EdgeClassifier.value
 
-            # Classifier can be None if no labels have been selected
-            if classifier is None or len(classifier.known_classes) < 2:
-                result[0] = np.zeros((len(edge_features_df),), dtype=np.float32)
-                return
+        # Classifier can be None if no labels have been selected
+        if len(getattr(classifier, "known_classes", [])) == 0:
```

Comment:
```suggestion
        if getattr(classifier, "known_classes", []):
```

I didn't suggest this before because IMO `if n == 0:` was better than `if not n:`.

Reply:
with `if n`.. I'll have to reverse the logic. I think I'll leave it with `len`.

Reply:
Now I see an obvious refactoring:

```python
known_classes = set(getattr(classifier, "known_classes", []))
if known_classes == {}:
    ...
elif known_classes == {1}:
    ...
...
```

Reply:
niiiice! :)

Reply:
side note: `{}` will be an empty dict and `!=set()`

Reply:
Another color for the bike shed:

```python
known_classes = sorted(getattr(classifier, "known_classes", []))
if known_classes in ([], [1]):
    ...
elif known_classes  == [2]:
    ...
elif known_classes == [1, 2]:
    ...
else:
    ...
```

This fails if `known_classes` has duplicates, which is probably good.

Reply:
`known_classes` should not have duplicates. :)